File: Algorithms/p104_Maximum_Depth_of_Binary_Tree_2.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    int maxDepth(TreeNode* root) {
        if(!root){
            return 0;
        }
        if(!root->left && !root->right){
            return 1;
        }
        int left_depth = 0;
        int right_depth = 0;
        if(root->left){
            left_depth = 1 + maxDepth(root->left);
        }
        if(root->right){
            right_depth = 1 + maxDepth(root->right);
        }
        return left_depth > right_depth ? left_depth : right_depth;
    }
};

File: Algorithms/p104_Maximum_Depth_of_Binary_Tree/p104_Maximum_Depth_of_Binary_Tree_2.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def maxDepth(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        if not root.left and not root.right:
            return 1
        left_depth = 0
        right_depth = 0
        if root.left:
            left_depth = 1 + self.maxDepth(root.left)
        if root.right:
            right_depth = 1 + self.maxDepth(root.right)
        return left_depth if left_depth > right_depth else right_depth
